torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -61,127 +63,903 @@ class SymbolicSizePrinter : private kir::IrVisitor {
+  }
+
+  const auto& allReads() {
+    return all_read_pos_;
+  }
+
+  auto firstWrite() const {
+    return first_write_pos_;
+  }
+
+  auto lastRead() const {
+    return last_read_pos_;
+  }
+
+  std::string toString() {
+    std::stringstream ss;
+    ss << "[ " << first_write_pos_ << " , " << last_read_pos_ << " )";
+    return ss.str();
+  }
+
+ private:
+  int first_write_pos_ = -1;
+  int last_read_pos_ = -1;
+  std::vector<int> all_read_pos_;
+};
+
+using BufferLiveIntervalPtrList = std::vector<BufferLiveInterval*>;
+
+//! Thin struct to keep track of loops. The actual loop body is
+//!  considered live in [start_pos, end_pos)
+struct ScopeInfo {
+  int start_pos = -1;
+  int end_pos = -1;
+
+  // nullptr means it's global scope
+  kir::ForLoop* loop = nullptr;
+};
+
+using ScopeInfoOwningPtr = std::unique_ptr<ScopeInfo>;
+using ScopeInfoOwningPtrList = std::vector<ScopeInfoOwningPtr>;
+
+//! Utility class to record the read and write of each
+//! allocated buffer.
 //!
-class AllocateReuseModifier {
-  // Alias local memory if it exceeds this threshold
-  static constexpr size_t kRegisterSizeThreshold = 1;
+//! Note:
+//!  this simplified interval analysis only works on pointwise ops and
+//!  reductions and broadcast. With no non-trivial IfThenElse and no
+//!  non-trivial re-computation.
+//!
+//!  Will probably at some point need dataflow and index analysis to precisely
+//!  handle loop carried dependency.
+struct AllocationUseDefInfo {
+  kir::Allocate* alloc_expr = nullptr;
+  kir::Allocate* alias_to = nullptr;
+  bool is_inner_alias = false;
+  bool should_try_alias = true;
+  MemoryType mem_type = MemoryType::Local;
+  DataType data_type = DataType::Float;
+  std::string size_expr;
+  ScopeInfo* loop_info = nullptr;
+  bool can_use_inner_alias = true;
+  int alloc_pos = -1;
+  std::unique_ptr<std::vector<AllocationUseDefInfo*>> inner_alias_list_ =
+      nullptr;
+  std::unique_ptr<BufferLiveInterval> inner_live_interval = nullptr;
+  std::unique_ptr<BufferLiveIntervalPtrList> inner_subscribed_intevals =
+      nullptr;
+  std::unique_ptr<BufferLiveInterval> outer_live_interval = nullptr;
+  std::unique_ptr<BufferLiveIntervalPtrList> outer_subscribed_intevals =
+      nullptr;
+};
 
+using AllocationInfoOwningPtr = std::unique_ptr<AllocationUseDefInfo>;
+using AllocationInfoOwningList = std::vector<AllocationInfoOwningPtr>;
+using AllocationInfoPtr = AllocationUseDefInfo*;
+using AllocationInfoList = std::vector<AllocationInfoPtr>;
+
+//! Analysis pass to collect the liveness info of local and shared buffers:
+//! The liveness info is illustrated as follows:
+//!
+//! For Idx0 ...
+//!   Alloc(T1, register)
+//!   Alloc(T2, register)
+//!   Alloc(T3, register)
+//!
+//!   For Idx1 ...     <---------- Outer Live Interval of T1 begin
+//!     For Idx2 ...
+//!       T1 = ...            <--  Inner Live Interval of T1 begin
+//!       T2 = ...
+//!       T3 = T1 + ...    <-- Inner Live Interval of T1 end
```

Comment:
Yes exactly for the outer loop part. 

For the inner liveness range, it will be extended to the line-number of the last expression using T1 in the other "outer" loop. 

Inner interval is meant to detect in-place sharing opportunities, so the pass uses most aggressive line-number requirement and relies on checks on producer-consumer relationship and index map to validate.